Mathlib/Tactic/Relation/Rfl.lean

```diff
@@ -38,22 +39,28 @@ initialize registerBuiltinAttribute {
     reflExt.add (decl, key) kind
 }
 
-open Elab.Tactic in
-/--
+/-- `MetaM` version of the `rfl` tactic.
+
 This tactic applies to a goal whose target has the form `x ~ x`, where `~` is a reflexive
 relation, that is, a relation which has a reflexive lemma tagged with the attribute [refl].
 -/
-elab_rules : tactic
-| `(tactic| rfl) => withMainContext do
-  let tgt ← getMainTarget
-  let .app (.app rel _) _ := tgt
-    | throwError "reflexivity lemmas only apply to binary relations, not {indentExpr tgt}"
+def _root_.Lean.MVarId.rfl (goal : MVarId) : MetaM Unit := do
+  let .app (.app rel _) _ ← whnfR <|← instantiateMVars <|← goal.getType
+    | throwError "reflexivity lemmas only apply to binary relations, not
+      {indentExpr (← goal.getType)}"
   let s ← saveState
   for lem in ← (reflExt.getState (← getEnv)).getMatch rel do
     try
-      liftMetaTactic (·.apply (← mkConstWithFreshMVarLevels lem))
-      return
+      let _ ← goal.apply (← mkConstWithFreshMVarLevels lem)
```

Comment:
```suggestion
      let gs ← goal.apply (← mkConstWithFreshMVarLevels lem)
      unless gs.isEmpty do
        logError <| MessageData.tagged `Tactic.unsolvedGoals <| m!"unsolved goals\n{goalsToMessageData gs}"
```
Ideally you would use `Term.reportUnsolvedGoals` here, but it is in `TermElabM` for no apparent reason.